stdlib/public/Synchronization/Mutex/LinuxImpl.swift

```diff
@@ -0,0 +1,150 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Atomics open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SynchronizationShims
+
+extension Atomic where Value == UInt32 {
+  borrowing func wait() -> Bool {
+    _swift_stdlib_wait(.init(rawAddress))
+  }
+
+  borrowing func wake() {
+    _swift_stdlib_wake(.init(rawAddress))
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+@frozen
+@usableFromInline
+@_staticExclusiveOnly
+internal struct _MutexHandle: ~Copyable {
+  // There are only 3 different values that storage can hold at a single time.
+  // 0: unlocked
+  // TID: locked, current thread's id (uncontended)
+  // (TID | SWIFT_FUTEX_WAITERS): locked, current thread's id (contended)
+  @usableFromInline
+  let storage: Atomic<UInt32>
+
+  @available(SwiftStdlib 5.11, *)
+  @_alwaysEmitIntoClient
+  @_transparent
+  init() {
+    storage = Atomic(0)
+  }
+
+  @available(SwiftStdlib 5.11, *)
+  @usableFromInline
+  borrowing func lock() {
+    // TODO: Is it worth caching this value in TLS?
+    let selfId = _swift_stdlib_gettid()
```

Comment:
It probably _is_ worth caching that, yes, because `gettid()` is a syscall, and part of the point of the other code here is to avoid doing a syscall unless we have to.